internal/mode/static/state/graph/graph.go

```diff
@@ -99,6 +103,14 @@ func (g *Graph) IsReferenced(resourceType client.Object, nsname types.Namespaced
 		// Service Namespace should be the same Namespace as the EndpointSlice
 		_, exists := g.ReferencedServices[types.NamespacedName{Namespace: nsname.Namespace, Name: svcName}]
 		return exists
+	// Similar to Namespace above, NginxProxy reference exists if it once was or currently is linked to a GatewayClass.
+	case *ngfAPI.NginxProxy:
+		var existed bool
+		if g.NginxProxy != nil {
+			existed = client.ObjectKeyFromObject(obj) == client.ObjectKeyFromObject(g.NginxProxy)
+		}
+		exists := isNginxProxyReferenced(obj, g.GatewayClass)
```

Comment:
`IsReferenced` is called before the graph is built for this particular update event, so it may not yet be there.

Reply:
IsReferenced is called on the last graph... so g.NginxProxy will be set if g.GatewayClass references an NginxProxy that exists in the last graph.

Reply:
This is for the case where NginxProxy wasn't referenced before (not in paramsRef), and then gets updated or added to be referenced.

Reply:
These are the two cases that trigger a change:
- GatewayClass references np-1, but np-1 doesn't exist so g.NginxProxy is nil. Then, np-1 is created.
- GatewayClass references np-1, and np-1 exists, so g.NginxProxy == np-1. Np-1 is updated or deleted. 

I think both these cases are covered by `isNginxProxyReferenced(obj, g.GatewayClass)`, right?

Reply:
Yes, I've gotten myself confused now. I was considering the case where the GC doesn't reference it, then it does. But that wouldn't even hit this code path. BuildGraph would properly add it to the Graph. So you're right.

Reply:
This method always confuses me. That's why I had to write out the cases haha

Reply:
`IsReferenced` determines if we upsert/remove the resource in our store, right? So what about the case where the GC has its paramsRef removed. Then we get an update to np-1, and since it's not linked to the GC, we don't process it. Then it gets added back to the GC, but in our store it's out of date.

Though this isn't fixed by my current logic because np-1 would have been removed from the Graph, so IsReferenced still returns false.

Reply:
This may mean we just have to keep all NginxProxies in our store, since they are first class resources that lead to NGINX config. Don't we keep all HTTPRoutes in our store, even if they don't yet reference a Gateway?

@pleshakov

Reply:
> IsReferenced determines if we upsert/remove the resource in our store, right? 

I believe store always persist unless we use nil in ChangeProcessor:
```
			{
				gvk:       extractGVK(&discoveryV1.EndpointSlice{}),
				store:     nil,
				predicate: funcPredicate{stateChanged: isReferenced},
			},
```
https://github.com/nginxinc/nginx-gateway-fabric/blob/b4b0772697d550314769519037acd4143ef98ca5/internal/mode/static/state/store.go#L184-L188

so it is separate from isRefernced, which is a predicate.

Reply:
> So does the nil predicate just determine if we build a new config? If so, am I right in saying that we build new config for any HTTPRoute event, even if irrelevant?

yep, looks like @kate-osborn  is working on it https://github.com/nginxinc/nginx-gateway-fabric/issues/1123 ;)